File: tg-core/infrastructure/secret_box.py
"""AES-GCM secret format shared with tg-web: v1.<b64-iv>.<b64-ciphertext>."""

from __future__ import annotations

import base64
import binascii
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

SECRET_BOX_VERSION = "v1"
LLM_API_KEY_AAD = "llm-provider-api-key"


def load_master_key_from_env(
    env_name: str = "BILLING_CONFIG_ENCRYPTION_KEY",
) -> bytes | None:
    raw = os.environ.get(env_name, "").strip()
    if not raw:
        return None
    try:
        key = _decode_base64(raw)
    except (binascii.Error, ValueError) as exc:
        raise ValueError(f"{env_name} must be Base64-encoded") from exc
    if len(key) != 32:
        raise ValueError(f"{env_name} must decode to exactly 32 bytes")
    return key


def decrypt_secret(value: str, context: str, key: bytes) -> str:
    version, encoded_iv, encoded_ciphertext = value.split(".", 2)
    if version != SECRET_BOX_VERSION or not encoded_iv or not encoded_ciphertext:
        raise ValueError("Unsupported encrypted secret format")
    plaintext = AESGCM(key).decrypt(
        _decode_base64(encoded_iv),
        _decode_base64(encoded_ciphertext),
        context.encode("utf-8"),
    )
    return plaintext.decode("utf-8")


def decrypt_llm_api_key(ciphertext: str, key: bytes | None = None) -> str:
    master = key if key is not None else load_master_key_from_env()
    if master is None:
        raise ValueError(
            "BILLING_CONFIG_ENCRYPTION_KEY is required to decrypt LLM API keys"
        )
    return decrypt_secret(ciphertext, LLM_API_KEY_AAD, master)


def _decode_base64(value: str) -> bytes:
    normalized = value.replace("-", "+").replace("_", "/")
    padded = normalized + ("=" * ((4 - len(normalized) % 4) % 4))
    return base64.b64decode(padded)

File: tg-web/src/backend/security/secret-box.ts
/**
 * AES-GCM 密文格式，与计费配置共用：`v1.<base64-iv>.<base64-ciphertext>`。
 * AAD（context）必须在加解密时一致。
 */
const encoder = new TextEncoder();
const decoder = new TextDecoder();
export const SECRET_BOX_VERSION = 'v1';

export function isValidSecretBoxMasterKey(value: string): boolean {
  try {
    return decodeBase64(value).byteLength === 32;
  } catch {
    return false;
  }
}

export async function importSecretBoxMasterKey(encodedMasterKey: string) {
  const bytes = decodeBase64(encodedMasterKey);
  if (bytes.byteLength !== 32) {
    throw new Error(
      'Encryption key must be a Base64-encoded 32-byte key',
    );
  }
  return crypto.subtle.importKey('raw', bytes, 'AES-GCM', false, [
    'encrypt',
    'decrypt',
  ]);
}

export async function encryptSecret(
  plaintext: string,
  context: string,
  key: CryptoKey | Promise<CryptoKey>,
) {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const ciphertext = await crypto.subtle.encrypt(
    {
      name: 'AES-GCM',
      iv,
      additionalData: encoder.encode(context),
    },
    await key,
    encoder.encode(plaintext),
  );
  return [SECRET_BOX_VERSION, encodeBase64(iv), encodeBase64(ciphertext)].join(
    '.',
  );
}

export async function decryptSecret(
  value: string,
  context: string,
  key: CryptoKey | Promise<CryptoKey>,
) {
  const [version, encodedIv, encodedCiphertext] = value.split('.');
  if (version !== SECRET_BOX_VERSION || !encodedIv || !encodedCiphertext) {
    throw new Error('Unsupported encrypted secret format');
  }
  const plaintext = await crypto.subtle.decrypt(
    {
      name: 'AES-GCM',
      iv: decodeBase64(encodedIv),
      additionalData: encoder.encode(context),
    },
    await key,
    decodeBase64(encodedCiphertext),
  );
  return decoder.decode(plaintext);
}

export function maskSecretHint(value: string): string {
  const trimmed = value.trim();
  if (trimmed.length <= 8) return '••••';
  return `${trimmed.slice(0, 4)}...${trimmed.slice(-4)}`;
}

function encodeBase64(value: ArrayBuffer | Uint8Array) {
  const bytes = value instanceof Uint8Array ? value : new Uint8Array(value);
  let binary = '';
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

function decodeBase64(value: string) {
  const normalized = value.replaceAll('-', '+').replaceAll('_', '/');
  const padded = normalized.padEnd(
    normalized.length + ((4 - (normalized.length % 4)) % 4),
    '=',
  );
  const binary = atob(padded);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0));
}
